### DISCUSSION TITLE
Custom composer extension method is not used

### DISCUSSION THREAD

**Discussion by Sebosek** [AutoFixture/AutoFixture#1227] [Q&A] [answered]
Hi,

at first, sorry for a little bit misleading title, but in fact I want to do very simple thing. I'm using AutoFixture for generating random data for stress test tool and I want to define value factory for each DTO property.
I've created extension methods like `AddRandomCountry()`

```csharp
public static IPostprocessComposer<T> AddRandomCountry<T>(this ICustomizationComposer<T> composer)
    where T : Order
{
    composer.With(p => p.Country, () =>
        Country[Random.Next(0, Country.Length)]);
    
    return composer;
}
```

With straight-forward usage, however created `order` instance contains "Country{GUID}" value instead of the one of the country codes. If I replace `AddRandomCountry()` method chain with `.With(p => p.Country, () => "Foo")`, the value "Foo" is presented.
I guess, there will be some kind of type validation, so question is: what have I done wrong?

```csharp
var order = fixture.Build<Order>().AddRandomCountry().Create();
```

**Comment by Ergamon** [comment]
Someone with more insight in the implementation has to answer why this does not work.

Do you really have subclasses of Order?

If not, what you want seems to be a simple customization:
```
      var fixture = new Fixture();
      fixture.Customize<Order>(c => c.With(o => o.Country, () => Country[Random.Next(0, Country.Length)]));
      var order = fixture.Create<Order>();
```

**Comment by Sebosek** [comment]
Hi @Ergamon, thanks for reply.  Sure, I could have all factories as part of the data generator, but from various reasons I don't want to. Personally, sooner or later I believe I would came across similar issue but from different reasons.
About the question of subclasses of Order; yes, I would like to have an extension method covering the Order class as a base class.

**Comment by aivascu** [comment, marked as answer]
@Sebosek first of all please use the **Discussions** section for questions and provide a minimum reproducible example.

The issues with your extension method is you're not returning the customized value. A way to imagine the postprocess composer is like a LINQ query builder against `IQueryable`. Each `With` method creates a new instance (node) in containing all previous customizations. You have customized the value the way you wanted to but you returned the old customization node which did not know of the way you customized the `Country`.

Simply put your customization extension method should look like this:
```cs
public static IPostprocessComposer<T> AddRandomCountry<T>(this ICustomizationComposer<T> composer)
    where T : Order
{
     return composer.With(p => p.Country, () =>
        Country[Random.Next(0, Country.Length)]);
}
```

See [here](https://dotnetfiddle.net/AZiB5u) a working example.

Moving to discussions
